[Uptime] Add throttling parameters to the Fleet UI

lucasfcosta opened the pull request:
## Summary

> ⚠️ This PR depends on the integration in elastic/integrations#2161 for its functional tests to pass on CI. Please merge that PR first.

This PR closes elastic/kibana#114155 by adding throttling parameters to the Fleet UI. These parameters are: `download`, `upload`, and `latency`. Furthermore, there's a field which isn't shown called `throttling.config`, which will merge the formatted `download`, `upload`, and `latency` values into one configuration parameter to be passed to the Synthetics runner.

![throttle_toggle](https://user-images.githubusercontent.com/6868147/143488004-7bd83b33-28e5-4d8c-abfb-3beee3783401.gif)

It includes:
* Validating that `download` and `upload` fields are greater than zero (when filled).
* Validating that `latency` is equal to or greater than zero (when filled).


## How to test this PR

**⚠️ NOTE: You'll need to install the `synthetics` integration version in elastic/integrations#2161 to be able to test this. To do so, you can use `elastic-package` to `build` it, and `install` it, as per [the instructions on their repo](https://github.com/elastic/elastic-package).**

This PR is somewhat tricky to test because you'll need to measure the network speed accurately on the agent itself. Therefore, you'll have to access websites which measure network speed, and for, latency, you'll have to use the [`NetworkInformation` API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation) to obtain the `rtt` (round-trip-time).

### Test Case 1 - Download and Upload Speeds:

1. Setup an Elastic Stack which has an Elastic Agent complete enrolled to a policy so that you can run browser monitors
2. Create a new browser monitor which will access `fast.com` and run the speed tests there. Make sure to give at least 30 seconds to run, and set its interval to 45 seconds. For that monitor, you can use the following inline script:
    ```js
    step("load homepage", async () => {
        await page.goto('https://www.fast.com');
    });
    
    step("wait for test", async () => {
      await page.waitForSelector('#show-more-details-link');
      await page.waitForTimeout(10000);
      await page.click("#show-more-details-link");
    });
    
    step("wait for up/lat", async () => {
      await page.waitForSelector(".extra-measurement-value-container.succeeded");
      await page.waitForTimeout(10000);
    });
    ```
    **Before saving the monitor, set the download and upload speeds to values which are smaller than your own upload/download speed, so that you can actually see the capped values in your test.**
3. Access Uptime > Monitors and go to your monitor's details.
4. Verify the monitor's screenshots to confirm that the download and upload speeds match your configured Latency value.
    ![Speed](https://user-images.githubusercontent.com/6868147/142046615-2d8eab21-1744-4e67-aa07-41466ad0756b.png)


### Test Case 2 - Latency:

1. Create a new folder which contains an `index.html` document with the following content:
    ```html
    <!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <title>Page Test</title>
    </head>
    <body>

        <h1>RTT: <span id="rtt-value">None</span></h1>

        <script type="application/javascript">
            const update = (id, val ) => document.getElementById(id).innerHTML = val

            const logNetworkInfo = () =>  {
                update('rtt-value', navigator.connection.rtt);
            }

            navigator.connection.addEventListener('change', logNetworkInfo);
            logNetworkInfo();
        </script>
    </body>
    </html>
    ```
2. Serve that `index.html` through a simple HTTP server using `npx http-server ./example_folder` (where `example_folder` is the folder with the `index.html` file from the previous step).
3. Setup an Elastic Stack which has an Elastic Agent complete enrolled to a policy so that you can run browser monitors
4. Create a new browser monitor which will access `host.docker.internal:8081` (that's the name Docker resolves to your `localhost`) so that you can see your `RTT` measurement.
    ```js
    step("load homepage", async () => {
        await page.goto('http://host.docker.internal:8081');
    });

    step("wait for test", async () => {
      await page.waitForTimeout(2000);
    });
    ```
    On that monitor **set the `Latency` field to the value you want to test** (I'd recommend `0`, `250` and `500`).
5. Access Uptime > Monitors and go to your monitor's details.
6. Verify the monitor's screenshots to confirm that RTT matches your configured Latency value.
    **NOTE**: The `rtt` property in the `NetworkInformation` API will round values to the nearest multiple of 25ms, as described [here](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt), so I recommend testing it with values like `0`, `250`, and `500` to see if the value displayed does mach the configured Latency on a 25-50ms range.
    ![rtt](https://user-images.githubusercontent.com/6868147/142046554-ab54479a-9104-4d6e-9cd3-4cf31f841b2c.png)



### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials - @dominiqueclarke is there any docs to update in this case? I couldn't find any relevant docs to update and didn't see any on [a similar PR you've done before](https://github.com/elastic/kibana/pull/112454/files).
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

| Risk                                                                          | Probability | Severity | Mitigation/Notes                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|-------------------------------------------------------------------------------|-------------|----------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| The handling of default values changes in the underlying `synthetics` module. | Low         | Medium   | Currently, the underlying `synthetics` module takes a single value with `/` separated fields for each of the throttling values. When one of the fields is not present, it uses default values for it. If we change the formatting of that flag's value or the behaviour of empty values, we will alter the behaviour of the throttling feature. To mitigate this, we should add tests for those flags, or separate them into multiple flags when running the underlying `synthetics` module. **Input on this would be appreciated**. |


### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


## Release Note

Allows users to set Download Speed, Upload Speed, and Latency for their synthetic monitors.

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

dominiqueclarke commented on x-pack/plugins/uptime/public/components/fleet_package/types.tsx:
```diff
@@ -115,6 +115,9 @@ export enum ConfigKeys {
   TIMEOUT = 'timeout',
+  DOWNLOAD_SPEED = 'throttling.download_speed',
+  UPLOAD_SPEED = 'throttling.upload_speed',
+  LATENCY = 'throttling.latency',
```
We may want to rethink putting this into 3 values. Here's why:

This works well for the Elastic Synthetics integration with Fleet. We have multiple layers of formatting to format fields in a way that fleet and the heartbeat will understand. The formatting flow is.

Raw UI format (the one we use for the components) -> UI formatters -> fleet -> package `${datastream}.yml.hbs` file.

Right now the `.yml.hbs` file is doing the work of formatting this into viable heartbeat.yml format. However, in the new UI monitor management flow we won't have that additional formatting layer.

We can add one, of course, but it's something to consider. In the past, we've made configuration keys that logically are multiple fields to a user, but one field in heartbeat, an object with multiple keys. We've then used our UI formatters to format that into one value to pass to fleet. 

I thought that this choice worked well when we were just considering fleet. But now that we'll be interacting with heartbeat directly, we'll either need to make this one config key that we store as an object in the UI but format as a string in our formatters, or we need to consider adding another layer of formatting to remove unused fields or merge fields.

lucasfcosta replied:
That's an excellent suggestion. I 100% agree.

Considering we'll now be interacting with `heartbeat` directly it would indeed be good to have one fewer layer of formatting for these values.

To be philosophical:

> _Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away._
> _Antoine de Saint-Exupéry, Airman's Odyssey_

I'll update the PR right now :)

dominiqueclarke reviewed (commented):
Code looks good to me, with one comment. I did not get a chance to review functionality before leaving, but I'll do a dep dive on Monday.

lucasfcosta commented:
@elasticmachine merge upstream

dominiqueclarke reviewed (changes requested):
I have begun testing this. Upload and download throttling looks great.

However, I am confused about the steps for the latency testing. Pinged on slack for more information. 

One thing that I think would be helpful is including default values for each field. This will give the user more context on what the browser monitors are run with by default, and give them a general baseline to adjust from.

We similarly fill in defaults for things like schedule, or screenshots seen here.
<img width="713" alt="Screen Shot 2021-11-22 at 3 41 51 PM" src="https://user-images.githubusercontent.com/11356435/142932439-3a166870-a74c-4f0e-85c1-57c708dedeaa.png">

dominiqueclarke reviewed (changes requested):
Default and reduced throttling config looks good, as well as RTT

<img width="1294" alt="Screen Shot 2021-11-23 at 10 22 38 PM" src="https://user-images.githubusercontent.com/11356435/143170371-3ad86689-8725-4417-82c3-eb2667d024cf.png">
<img width="1307" alt="Screen Shot 2021-11-23 at 10 18 40 PM" src="https://user-images.githubusercontent.com/11356435/143170373-6bcd8a20-aeeb-4379-bfc5-39d3bda71402.png">
<img width="1299" alt="Screen Shot 2021-11-23 at 10 36 37 PM" src="https://user-images.githubusercontent.com/11356435/143170482-b043a3e3-aaea-4c53-985f-7ded9515744c.png">

However, as I'm looking through it more as a user, I'm realizing this ticket does not satisfy all of the AC. 
The AC also states:
- Option to disabling throttling

The documentation states that throttling can be disabled by passing `throttling: false`. 

Therefore, I think we need a checkbox or toggle to enable or disable throttling. The toggle or checkbox should be set on by default, with the default throttling values from heartbeat. When unchecked, it should pass false as the value for the `throttling.config` key.

lucasfcosta commented:
@dominiqueclarke I've just updated this PR with the `Switch` toggle. As I was doing that, I came across a few things I would like your input on:

* I've made the enable/disable toggle its own flag and separated fields into another component so that the code is more readable and so that we can test it in a separate suite. 
    * Should we merge components?
* I've used a formatter to conditionally set the throttling field which gets passed to the agent as `false`. I did that because of your recommendation to avoid formatting within the handlebars file. Now we don't have to do any formatting there at all and everything is passed exactly as it will be used by the agent. What do you think about this approach?
* Given I've added another `EuiSwitch`, some tests started failing because the test environment from `EUI` will generate the same `aria-labelledby` attribute for all switches, thus causing `react-testing-library` to find _two_ elements matching the label query. See this behaviour [here](https://github.com/elastic/eui/blob/91b416dcd51e116edb2cb4a2cac4c306512e28c7/src/services/accessibility/html_id_generator.testenv.ts#L12).

On another note, here's how the new version of this feature looks like:

![throttle_toggle](https://user-images.githubusercontent.com/6868147/143488004-7bd83b33-28e5-4d8c-abfb-3beee3783401.gif)

This also addresses @shahzad31's point about the repetitive text.

shahzad31 commented:
> @dominiqueclarke I've just updated this PR with the `Switch` toggle. As I was doing that, I came across a few things I would like your input on:
> 
> * I've made the enable/disable toggle its own flag and separated fields into another component so that the code is more readable and so that we can test it in a separate suite. 
>     * Should we merge components?
> * I've used a formatter to conditionally set the throttling field which gets passed to the agent as `false`. I did that because of your recommendation to avoid formatting within the handlebars file. Now we don't have to do any formatting there at all and everything is passed exactly as it will be used by the agent. What do you think about this approach?
> * Given I've added another `EuiSwitch`, some tests started failing because the test environment from `EUI` will generate the same `aria-labelledby` attribute for all switches, thus causing `react-testing-library` to find _two_ elements matching the label query. See this behaviour [here](https://github.com/elastic/eui/blob/91b416dcd51e116edb2cb4a2cac4c306512e28c7/src/services/accessibility/html_id_generator.testenv.ts#L12).
> 
> On another note, here's how the new version of this feature looks like:
> 
> ![throttle_toggle](https://user-images.githubusercontent.com/6868147/143488004-7bd83b33-28e5-4d8c-abfb-3beee3783401.gif)
> 
> This also addresses @shahzad31's point about the repetitive text.
> 

Looks much better, though just to reduce the noise and encourage default usage.


I propose that instead of a toggle Enable throttling

We should add a toggle like this

"Use default throttling"
Add a small description of what default throttling is
"5mbps download, 10 upload, 20 latency"

Be default it will be enabled and we hide all the related text fields and if user disables it ,in that case we allow them to make changes via text boxes.


In subsequent renders while editing, we can determine if they have changed default values to determine the toggle state.


Again consider this just a suggestion :)

shahzad31 commented:
Thinking again, my proposed solution doesn't consider how user with disable throttling so you can probably discard that.

dominiqueclarke commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

juliaElastic reviewed (approved):
minor change in fleet, LGTM

dominiqueclarke reviewed (approved):
LGTM. Editing policies is working as expected thanks to the normalizer.

lucasfcosta commented:
Merging as we've had "through the eyes of the user" tests, design review, and technical review, including E2E tests on all of them. Product review is not available atm given personal circumstances. If they'd like to to review this too I'd be happy to add further changes though, of course.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 118594` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.0](https://github.com/elastic/kibana/pull/120502)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/120502">](https://github.com/elastic/kibana/pull/120502) |

This backport PR will be merged automatically after passing CI.
